src/composite.c

```diff
@@ -0,0 +1,195 @@
+
+  // These images will all be unreffed when context is unreffed:
+  VipsImage **t = (VipsImage **) vips_object_local_array( context, 3 );
+
+  VipsImage *srcRGB;
+  VipsImage *srcAlpha;
+  VipsImage *srcAlphaNormalized;
+
+  VipsImage *dstRGB;
+  VipsImage *dstRGBPremultiplied;
+  VipsImage *dstAlpha;
+  VipsImage *dstAlphaNormalized;
+
+  VipsImage *outRGB;
+  VipsImage *outRGBPremultiplied;
+  VipsImage *outAlpha;
+  VipsImage *outAlphaNormalized;
+
+  // Extract RGB bands
+  if( vips_extract_band( src, &srcRGB, 0, "n", NUM_COLOR_BANDS, NULL ) ||
+      vips_extract_band( dst, &dstRGB, 0, "n", NUM_COLOR_BANDS, NULL ) )
+    return( -1 );
+
+  // Extract alpha bands
+  if( vips_extract_band( src, &srcAlpha, ALPHA_BAND_INDEX, NULL ) ||
+      vips_extract_band( dst, &dstAlpha, ALPHA_BAND_INDEX, NULL ) )
+    return( -1 );
+
+  // Compute normalized input alpha channels:
+  if( vips_linear1( srcAlpha, &srcAlphaNormalized, 1.0 / 255.0, 0.0, NULL ) ||
+      vips_linear1( dstAlpha, &dstAlphaNormalized, 1.0 / 255.0, 0.0, NULL ) )
+    return( -1 );
+
+  //
+  // Compute normalized output alpha channel:
+  //
+  // References:
+  // - http://en.wikipedia.org/wiki/Alpha_compositing#Alpha_blending
+  // - https://github.com/jcupitt/ruby-vips/issues/28#issuecomment-9014826
+  //
+  // out_a = src_a + dst_a * (1 - src_a)
+  //                         ^^^^^^^^^^^
+  //                            t[0]
+  //                 ^^^^^^^^^^^^^^^^^^^
+  //                        t[1]
+  if( vips_linear1( srcAlphaNormalized, &t[0], -1.0, 1.0, NULL ) ||
+      vips_multiply( dstAlphaNormalized, t[0], &t[1], NULL ) ||
+      vips_add( srcAlphaNormalized, t[1], &outAlphaNormalized, NULL ) )
+    return( -1 );
+
+  //
+  // Compute output RGB channels:
+  //
+  // Wikipedia:
+  // out_rgb = (src_rgb * src_a + dst_rgb * dst_a * (1 - src_a)) / out_a
+  //
+  // `vips_ifthenelse` with `blend=TRUE`: http://bit.ly/1KoSsga
+  // out = (cond / 255) * in1 + (1 - cond / 255) * in2
+  //
+  // Substitutions:
+  //
+  //     cond --> src_a
+  //     in1 --> src_rgb
+  //     in2 --> dst_rgb * dst_a (premultiplied destination RGB)
+  //
+  // Finally, manually divide by `out_a` to unpremultiply the RGB channels.
+  // Failing to do so results in darker than expected output with low
+  // opacity images.
+  //
+  if( vips_multiply( dstRGB, dstAlphaNormalized, &dstRGBPremultiplied, NULL ) )
+    return( -1 );
+
+  if( vips_ifthenelse( srcAlpha, srcRGB, dstRGBPremultiplied,
+      &outRGBPremultiplied, "blend", TRUE, NULL ) )
+    return( -1 );
+
+  // Unpremultiply RGB channels:
+  if( vips_divide( outRGBPremultiplied, outAlphaNormalized, &outRGB, NULL ) )
+    return( -1 );
+
+  // Denormalize output alpha channel:
+  if( vips_linear1( outAlphaNormalized, &outAlpha, 255.0, 0.0, NULL ) )
+    return( -1 );
+
+  // Combine RGB and alpha channel into output image:
+  if( vips_bandjoin2( outRGB, outAlpha, &t[2], NULL ) )
+    return( -1 );
+
+  // Source
+  g_object_unref( srcRGB );
```

Comment:
Thanks for the tip, will make that change :+1: I am still getting used to manual memory management :wink: